Src/Common/FSecure/C3/Interfaces/Channels/Mattermost.cpp

```diff
@@ -0,0 +1,252 @@
+			auto sendData = data.SubString(0, actualPacketSize);
+			auto foo = m_MattermostObj.WriteReply(cppcodec::base64_rfc4648::encode(sendData.data(), sendData.size()), postID);
+
+            // Error. Try to retransmit that packet.
+            if (foo.empty())
+                return 0;
+		}
+
+		//Update the original first message with "C2S||S2C:Done" - these messages will always be read in onRecieve.
+		std::string message = m_outboundDirectionName + OBF(":Done");
+
+		m_MattermostObj.UpdatePost(message, postID);
+		return actualPacketSize;
+	}
+	catch (...)
+	{
+        // Should exception be thrown, return 0 to make C3 retransmit that packet.
+		return 0;
+	}
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+std::vector<FSecure::ByteVector> FSecure::C3::Interfaces::Channels::Mattermost::OnReceiveFromChannel()
+{
+	auto messages = m_MattermostObj.GetMessagesByDirection(m_inboundDirectionName + OBF(":Done"));
+
+	std::vector<ByteVector> ret;
+
+	//Read the messages in reverse order (which is actually from the oldest to newest)
+	//Avoids old messages being left behind.
+	//for (std::vector<std::string>::reverse_iterator postID = messages.rbegin(); postID != messages.rend(); ++postID)
+	for (std::vector<std::string>::iterator postID = messages.begin(); postID != messages.end(); ++postID)
+	{
+		auto replies = m_MattermostObj.ReadReplies(*postID);
+		std::vector<std::string> postIDs;
+		std::string message;
+
+		//Get all of the messages from the replies.
+		for (auto&& reply : replies)
+		{
+			message.append(reply.second);
+			postIDs.push_back(std::move(reply.first)); //get all of the post_ids for later deletion
+		}
+		
+        auto relayMsg = cppcodec::base64_rfc4648::decode(message);
+		ret.emplace_back(std::move(relayMsg));
+
+		DeleteReplies(postIDs);
+		m_MattermostObj.DeletePost(*postID);
+	}
+
+	return ret;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+void FSecure::C3::Interfaces::Channels::Mattermost::DeleteReplies(std::vector<std::string> const & postIDs)
+{
+	for (auto&& postID : postIDs)
+	{
+		m_MattermostObj.DeletePost(postID);
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+FSecure::ByteVector FSecure::C3::Interfaces::Channels::Mattermost::OnRunCommand(ByteView command)
+{
+    auto commandCopy = command;
+    switch (command.Read<uint16_t>())
+    {
+    case 0:
+		m_MattermostObj.PurgeChannel();
+        return {};
+    case 1:
+        m_MattermostObj.SetToken(commandCopy.Read<std::string>());
```

Comment:
`commandCopy` is created so in default case, first two bytes of id will be passed to `AbstractChannel` that might have its own case to handle it. Here it will try to read a string and will use those two bytes as part of string size. `command` should be used instead.